libflash/optimized_writer.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include "optimized_writer.hpp"
+#include <iostream>
+
+using namespace mender;
+
+OptimizedWriter::OptimizedWriter(
+	io::FileReader &reader, io::FileReadWriterSeeker &writer, size_t blockSize, size_t volumeSize) :
+	mBlockSize(blockSize),
+	mReader(reader),
+	mReadWriter(writer),
+	mVolumeSize(volumeSize) {
+}
+
+Error OptimizedWriter::Copy(bool optimized) {
+	mStatistics.mBlocksWritten = 0;
+	mStatistics.mBlocksOmitted = 0;
+	mStatistics.mBytesWritten = 0;
+
+	io::Bytes rv;
+	rv.resize(mBlockSize);
+	io::Bytes wv;
+	wv.resize(mBlockSize);
+
+	while (true) {
+		if (rv.size() != mBlockSize) {
+			rv.resize(mBlockSize);
+		}
+
+		auto pos = mReader.Tell();
+		if (!pos) {
+			return pos.error();
+		}
+		auto position = pos.value();
+
+		if (mVolumeSize && ((position + mBlockSize) > mVolumeSize)) {
+			return NoError;
+		}
+
+		auto result = mReader.Read(rv.begin(), rv.end());
+		if (!result) {
+			return result.error();
+		} else if (result.value() == 0) {
+			return NoError;
+		} else if (result.value() > rv.size()) {
+			return mender::common::error::MakeError(
+				mender::common::error::ProgrammingError,
+				"Read returned more bytes than requested. This is a bug in the Read function.");
+		}
+
+		auto readBytes = result.value();
+		if (readBytes != rv.size()) {
+			// Because we only ever resize down, this should be very cheap. Resizing
+			// back up to capacity below is then also cheap.
+			rv.resize(readBytes);
+		}
+
+		if (wv.size() != readBytes) {
+			wv.resize(readBytes);
+		}
+
+		if (NoError != mReadWriter.SeekSet(position)) {
+			return Error(
+				std::error_condition(std::errc::io_error),
+				"Failed to set seek on the destination file");
+		}
+
+		bool skipWriting = false;
+		if (optimized) {
+			auto readResult = mReadWriter.Read(wv.begin(), wv.end());
+			if (readResult && readResult.value() == readBytes) {
```

Comment:
There are some unhandled cases here:

1. `readResult` error must be handled (this is just straight to error return).
2. `readResult.value() != readBytes` is not handled either. Currently what it will do is to write, since `skipWriting` will still be false. Which isn't wrong exactly, but it's silently doing the extra work that we don't want it to do, so I would suggest one of these approaches instead:
    1. Implement logic to `Read` multiple times so that we reach the expected value.
    2. Just return error. It's very unlikely that `Read` will in fact return a short value, so I think it's safe to just treat this as an error, and it's much easier than the first option.

Reply:
There is a "while" loop in, so it will read multiple times:
`ExpectedSize mender::io::Read(File f, uint8_t *dataPtr, size_t dataLen)`
It probably is very unlikely it will return a short value, but I can treat this as an error

Reply:
I implemented it like that because I was testing it on regular files. I had an input file and an empty output file, in that case it was just coping content on the input file to the output.

Reply:
We aren't really expecting to write to empty files, but still it is a good point. Maybe add a comment about `mender::io::Read` then, so that's it's clear that we are not expecting any short writes unless the file is empty.